src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,275 @@
+
+	foreach_ptr(usedSubPlanIdValue, usedSubPlanNodeList)
+	{
+		char *resultId = strVal(usedSubPlanIdValue);
+
+		IntermediateResultsHashEntry *entry = SearchIntermediateResult(
+			intermediateResultsHash, resultId);
+
+		/* no need to traverse the whole plan if all the workers are hit */
+		if (list_length(entry->nodeIdList) == workerNodeCount)
+		{
+			elog(DEBUG4, "Subplan %s is used in all workers", resultId);
+
+			break;
+		}
+		else
+		{
+			/*
+			 * traverse the plan and add find all worker nodes
+			 *
+			 * If we have reference tables in the distributed plan, all the
+			 * workers will be in the node list. We can improve intermediate result
+			 * pruning by deciding which reference table shard will be accessed earlier
+			 */
+			entry->nodeIdList = AppendAllAccessedWorkerNodes(entry->nodeIdList,
+															 distributedPlan,
+															 workerNodeCount);
+
+			elog(DEBUG4, "Subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultsHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered and
+ * all the workers with placements are appended to the list. This effectively
+ * means that if there is a reference table access in the distributed plan, all
+ * the workers will be in the resulting list.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan, int
+							 workerNodeCount)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = lfirst(taskCell);
+		ListCell *placementCell = NULL;
+		foreach(placementCell, task->taskPlacementList)
+		{
+			ShardPlacement *placement = lfirst(placementCell);
+			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
+
+			/* early return if all the workers are accessed */
+			if (list_length(workerNodeList) == workerNodeCount)
+			{
+				return workerNodeList;
+			}
+		}
+	}
+
+	return workerNodeList;
+}
+
+
+/*
+ * MakeIntermediateResultHTAB is a helper method that creates a Hash Table that
+ * stores information on the intermediate result.
+ */
+HTAB *
+MakeIntermediateResultHTAB()
+{
+	HTAB *intermediateResultsHash = NULL;
+	uint32 hashFlags = 0;
+	HASHCTL info = { 0 };
+	int initialNumberOfElements = 16;
+
+	info.keysize = sizeof(NAMEDATALEN);
```

Comment:
should be `= NAMEDATALEN;`

Reply:
No, It should have been `sizeof(char*)` here.
`= NAMEDATALEN;` ends up in backend crashes